### PR TITLE
Add index types to Vector class

### PR DESCRIPTION
**Pull request by DanielRX**
This fixes errors for 
```ts
const x = new Vector<string>('prefix');
x[0] = '';
``` 
as it will now type out the index

### CODE REVIEW

File: assembly/collections/vector.ts [hunk 1]
```diff
@@ -8,6 +8,9 @@ export class Vector<T> {
   private _elementPrefix: string;
   private _lengthKey: string;
   private _length: i32;
+
+  [key: string]: T;
```
Comment (janedegtiareva) [inline assembly/collections/vector.ts]: what's the intended usage of the string version?

File: assembly/collections/vector.ts [hunk 2]
```diff
@@ -8,6 +8,9 @@ export class Vector<T> {
...
+  [key: number]: T;
```
Comment (janedegtiareva) [inline assembly/collections/vector.ts]: Do you mind adding a test case in main.ts under tests?

File: assembly/collections/vector.ts (hunk 1 above)
Comment (DanielRX) [reply assembly/collections/vector.ts]: The string case was a catch all, for vector it is incorrect as I assumed `{}` as an operator was `x['key']`

File: assembly/collections/vector.ts (hunk 2 above)
Comment (DanielRX) [reply assembly/collections/vector.ts]: A test case for what? This just makes `tsc` happy with the index access to a vector (the string version is incorrect here)

File: assembly/collections/vector.ts (hunk 1 above)
Comment (willemneal) [reply assembly/collections/vector.ts]:
```ts
  @operator("{}")
  private __unchecked_get(index: i32): T {
```
`{}` is just the unchecked version of `[]`.

File: assembly/collections/vector.ts (hunk 2 above)
Comment (willemneal) [reply assembly/collections/vector.ts]: I guess if we want to enforce that tsc is happy with the source, just add `tsc --noEmit` as a test.

File: assembly/collections/vector.ts (hunk 1 above)
Comment (DanielRX) [reply assembly/collections/vector.ts]: Is that going to be used in file as `x{1}` or `x['1']`? i.e.: Is it going to be invalid typescript syntax or just string index keys?

**Pull request merged**